File: code/python/06-frameworks/langgraph_multi_agent.py
"""Multi-agent research-and-writing workflow built with LangGraph.

Workflow:
    START → research → fact_check → writer → editor
                                              ↓ (needs revision?)
                                           writer  ← loop
                                              ↓ (approved)
                                            END

Each agent is a specialised LLM node with specific instructions.
The editor decides whether to approve the report or send it back
for revision (up to MAX_REVISION_CYCLES times).

Run:
    python langgraph_multi_agent.py

See: docs/06-frameworks-in-practice/02-langchain-langgraph.md
"""

from __future__ import annotations

import json
import os
import time
from dataclasses import dataclass, field
from typing import Annotated, Optional

# ---------------------------------------------------------------------------
# Optional LangGraph imports (fail gracefully)
# ---------------------------------------------------------------------------

try:
    from typing import TypedDict

    from langchain_core.messages import HumanMessage, SystemMessage, AIMessage
    from langchain_openai import ChatOpenAI
    from langgraph.graph import StateGraph, END
    from langgraph.graph.message import add_messages

    LANGGRAPH_AVAILABLE = True
except ImportError:
    LANGGRAPH_AVAILABLE = False

MAX_REVISION_CYCLES = 3

# ---------------------------------------------------------------------------
# Shared state
# ---------------------------------------------------------------------------


class WorkflowState(TypedDict):
    """State shared across all nodes in the research-and-writing workflow."""

    topic: str
    research_notes: str          # Output from research agent
    verified_facts: str          # Output from fact-checker
    draft_report: str            # Output from writer
    editor_feedback: str         # Output from editor (if revision needed)
    approved: bool               # True when editor approves
    revision_count: int          # Number of write→edit cycles
    sources: list[str]           # Accumulated source references
    workflow_trace: list[dict]   # Step-by-step execution log


# ---------------------------------------------------------------------------
# Mock tool: web search
# ---------------------------------------------------------------------------

_SEARCH_MOCK: dict[str, list[dict]] = {
    "default": [
        {
            "title": "AI in Software Engineering 2026",
            "snippet": (
                "By 2026, AI coding assistants handle ~40% of routine code generation. "
                "GitHub Copilot, Cursor, and similar tools are used daily by 60% of professional developers. "
                "AI-assisted code review catches 35% more bugs than manual review alone."
            ),
            "url": "https://example.com/ai-software-2026",
        },
        {
            "title": "LLM-Driven Development Practices",
            "snippet": (
                "Test-driven development has evolved into prompt-driven development (PDD). "
                "Teams use LLMs to generate test cases, refactor legacy code, "
                "and produce API documentation automatically."
            ),
            "url": "https://example.com/pdd-2026",
        },
        {
            "title": "AI Agent Adoption in Engineering Teams",
            "snippet": (
                "Multi-agent systems now handle CI/CD pipeline configuration, "
                "security scanning, and performance profiling. "
                "84% of Fortune 500 tech teams report using some form of AI agent in their workflow."
            ),
            "url": "https://example.com/agent-adoption",
        },
    ]
}


def mock_web_search(query: str) -> list[dict]:
    """Simulate a web search. Returns mock results for demo purposes."""
    return _SEARCH_MOCK.get(query, _SEARCH_MOCK["default"])


# ---------------------------------------------------------------------------
# Result type
# ---------------------------------------------------------------------------


@dataclass
class WorkflowResult:
    """Structured output from :meth:`ResearchWritingWorkflow.run`."""

    report: str
    sources: list[str]
    revisions: int
    workflow_trace: list[dict]
    elapsed_ms: float


# ---------------------------------------------------------------------------
# ResearchWritingWorkflow
# ---------------------------------------------------------------------------


class ResearchWritingWorkflow:
    """Multi-agent research-and-writing workflow.

    Four specialised agents collaborate via LangGraph StateGraph:
      1. Research Agent   — gathers information, cites sources
      2. Fact-Checker     — verifies claims against research notes
      3. Writer Agent     — produces a structured Markdown report
      4. Editor Agent     — reviews quality; approves or sends back for revision

    Args:
        model:               OpenAI chat model for all agents.
        max_revision_cycles: Maximum writer → editor → writer loops.
    """

    def __init__(
        self,
        model: str = "gpt-4o",
        max_revision_cycles: int = MAX_REVISION_CYCLES,
    ) -> None:
        if not LANGGRAPH_AVAILABLE:
            raise ImportError(
                "LangGraph is required. Install with:\n"
                "  pip install langgraph langchain-openai"
            )
        self.model = model
        self.max_revision_cycles = max_revision_cycles
        self.llm = ChatOpenAI(model=model, temperature=0.3)
        self.graph = self._build_workflow()

    # ------------------------------------------------------------------
    # Agent builders
    # ------------------------------------------------------------------

    def _create_research_agent(self):
        """Return an LLM callable configured as a research agent."""
        system = SystemMessage(content=(
            "You are a research agent. Your job is to gather comprehensive, "
            "accurate information on a given topic. "
            "Simulate web searches and database queries. "
            "Always cite your sources with URLs. "
            "Output your findings as structured notes with clear sections. "
            "Format: use bullet points, be factual, cite everything."
        ))

        def research(state: WorkflowState) -> WorkflowState:
            topic = state["topic"]

            # Simulate web search results
            search_results = mock_web_search(topic)
            search_context = "\n".join(
                f"- [{r['title']}]({r['url']}): {r['snippet']}"
                for r in search_results
            )

            messages = [
                system,
                HumanMessage(content=(
                    f"Research this topic thoroughly: {topic!r}\n\n"
                    f"Available search results:\n{search_context}\n\n"
                    "Produce detailed research notes with key facts, statistics, "
                    "and source citations."
                )),
            ]
            response = self.llm.invoke(messages)
            notes = response.content

            sources = [r["url"] for r in search_results]
            trace_entry = {
                "step": "research",
                "agent": "ResearchAgent",
                "summary": f"Gathered {len(search_results)} sources on: {topic}",
                "output_preview": notes[:200] + "...",
            }

            return {
                **state,
                "research_notes": notes,
                "sources": state["sources"] + sources,
                "workflow_trace": state["workflow_trace"] + [trace_entry],
            }

        return research

    def _create_fact_checker(self):
        """Return an LLM callable configured as a fact-checker."""
        system = SystemMessage(content=(
            "You are a fact-checker. Your job is to verify claims in research notes. "
            "Cross-check each major claim. Flag any unsupported assertions. "
            "Output a verified facts summary, noting what was confirmed and what was uncertain."
        ))

        def fact_check(state: WorkflowState) -> WorkflowState:
            messages = [
                system,
                HumanMessage(content=(
                    f"Fact-check these research notes about: {state['topic']!r}\n\n"
                    f"{state['research_notes']}\n\n"
                    "Produce a verified facts summary. "
                    "Format: list confirmed facts, list uncertain claims."
                )),
            ]
            response = self.llm.invoke(messages)
            verified = response.content

            trace_entry = {
                "step": "fact_check",
                "agent": "FactChecker",
                "summary": "Verified research notes",
                "output_preview": verified[:200] + "...",
            }

            return {
                **state,
                "verified_facts": verified,
                "workflow_trace": state["workflow_trace"] + [trace_entry],
            }

        return fact_check

    def _create_writer(self):
        """Return an LLM callable configured as a writer."""
        system = SystemMessage(content=(
            "You are a technical writer. Produce clear, well-structured reports. "
            "Use Markdown formatting: headers, bullet points, and code blocks where relevant. "
            "Each factual claim must reference the research. "
            "If given editor feedback, address every point in your revision."
        ))

        def write(state: WorkflowState) -> WorkflowState:
            feedback_section = ""
            if state.get("editor_feedback"):
                feedback_section = (
                    f"\n\nEditor feedback to address:\n{state['editor_feedback']}\n"
                    "Please revise the report to address ALL points above."
                )
            revision_note = (
                f" (Revision {state['revision_count'] + 1})"
                if state["revision_count"] > 0
                else ""
            )

            messages = [
                system,
                HumanMessage(content=(
                    f"Write a comprehensive report{revision_note} on: {state['topic']!r}\n\n"
                    f"Research notes:\n{state['research_notes']}\n\n"
                    f"Verified facts:\n{state['verified_facts']}"
                    f"{feedback_section}\n\n"
                    "Produce a complete, publication-ready Markdown report."
                )),
            ]
            response = self.llm.invoke(messages)
            draft = response.content

            is_revision = state["revision_count"] > 0
            trace_entry = {
                "step": "write",
                "agent": "Writer",
                "summary": (
                    f"Wrote revision {state['revision_count'] + 1}"
                    if is_revision
                    else "Wrote initial draft"
                ),
                "output_preview": draft[:200] + "...",
            }

            return {
                **state,
                "draft_report": draft,
                "revision_count": state["revision_count"] + 1,
                "workflow_trace": state["workflow_trace"] + [trace_entry],
            }

        return write

    def _create_editor(self):
        """Return an LLM callable configured as an editor."""
        system = SystemMessage(content=(
            "You are a senior editor. Review reports for clarity, accuracy, completeness, "
            "and professional quality. "
            "If the report is ready to publish, respond with exactly: APPROVED\n"
            "followed by a brief note of what's good. "
            "If it needs work, respond with exactly: REVISION NEEDED\n"
            "followed by specific, actionable feedback. "
            "Be demanding but fair — only approve reports that meet publication standards."
        ))

        def edit(state: WorkflowState) -> WorkflowState:
            messages = [
                system,
                HumanMessage(content=(
                    f"Review this report about: {state['topic']!r}\n\n"
                    f"{state['draft_report']}\n\n"
                    f"This is revision {state['revision_count']} of {self.max_revision_cycles} allowed."
                )),
            ]
            response = self.llm.invoke(messages)
            feedback = response.content

            approved = feedback.strip().upper().startswith("APPROVED")

            # Force approval when max revisions reached
            if state["revision_count"] >= self.max_revision_cycles:
                approved = True
                feedback = f"APPROVED (max revisions reached)\n{feedback}"

            trace_entry = {
                "step": "edit",
                "agent": "Editor",
                "summary": "Approved report" if approved else "Requested revision",
                "output_preview": feedback[:200] + "...",
            }

            return {
                **state,
                "editor_feedback": feedback if not approved else "",
                "approved": approved,
                "workflow_trace": state["workflow_trace"] + [trace_entry],
            }

        return edit

    # ------------------------------------------------------------------
    # Graph assembly
    # ------------------------------------------------------------------

    def _build_workflow(self):
        """Declare the research-and-writing graph."""
        research_fn = self._create_research_agent()
        fact_check_fn = self._create_fact_checker()
        writer_fn = self._create_writer()
        editor_fn = self._create_editor()

        workflow = StateGraph(WorkflowState)

        workflow.add_node("research", research_fn)
        workflow.add_node("fact_check", fact_check_fn)
        workflow.add_node("writer", writer_fn)
        workflow.add_node("editor", editor_fn)

        workflow.set_entry_point("research")
        workflow.add_edge("research", "fact_check")
        workflow.add_edge("fact_check", "writer")

        workflow.add_conditional_edges(
            "editor",
            lambda state: "end" if state["approved"] else "writer",
            {"writer": "writer", "end": END},
        )
        workflow.add_edge("writer", "editor")

        return workflow.compile()

    @staticmethod
    def visualize() -> str:
        """Return an ASCII diagram of the workflow."""
        return """\
Multi-Agent Research-and-Writing Workflow
──────────────────────────────────────────

┌─────────┐
│  START  │
└────┬────┘
     ▼
┌──────────┐
│ research │  Gathers information, cites sources
└────┬─────┘
     ▼
┌────────────┐
│ fact_check │  Verifies claims
└─────┬──────┘
      ▼
┌────────┐
│ writer │ ◀──────────────────────────────────────────────────┐
└────┬───┘                                                     │
     ▼                                                         │
┌────────┐    REVISION NEEDED (revision_count < max)          │
│ editor │ ──────────────────────────────────────────────────▶ ┘
└────┬───┘
     │ APPROVED (or max revisions reached)
     ▼
┌─────────┐
│   END   │
└─────────┘
"""

    # ------------------------------------------------------------------
    # Public interface
    # ------------------------------------------------------------------

    def run(self, topic: str) -> WorkflowResult:
        """Research and write a report on *topic*.

        Args:
            topic: The research topic.

        Returns:
            :class:`WorkflowResult` with the final report, sources,
            revision count, and full workflow trace.
        """
        start = time.monotonic()

        initial_state: WorkflowState = {
            "topic": topic,
            "research_notes": "",
            "verified_facts": "",
            "draft_report": "",
            "editor_feedback": "",
            "approved": False,
            "revision_count": 0,
            "sources": [],
            "workflow_trace": [],
        }

        final_state = self.graph.invoke(initial_state)
        elapsed = (time.monotonic() - start) * 1000

        return WorkflowResult(
            report=final_state["draft_report"],
            sources=final_state["sources"],
            revisions=final_state["revision_count"],
            workflow_trace=final_state["workflow_trace"],
            elapsed_ms=elapsed,
        )


# ---------------------------------------------------------------------------
# Demo entry point
# ---------------------------------------------------------------------------


def print_workflow_trace(trace: list[dict]) -> None:
    """Pretty-print the workflow trace."""
    print("\nWorkflow Trace:")
    print("─" * 60)
    for i, entry in enumerate(trace, 1):
        agent = entry.get("agent", "unknown")
        summary = entry.get("summary", "")
        preview = entry.get("output_preview", "")
        print(f"  Step {i} [{agent}]: {summary}")
        print(f"    Preview: {preview[:100]}...")
        print()


def main() -> None:
    """Run the demo on a real research topic."""
    print(ResearchWritingWorkflow.visualize())

    if not LANGGRAPH_AVAILABLE:
        print("LangGraph is not installed. Install with:")
        print("  pip install langgraph langchain-openai")
        return

    topic = "The impact of AI on software engineering in 2026"
    print(f"Topic: {topic!r}\n")
    print("Running multi-agent workflow...")
    print("(This makes real LLM calls — ensure OPENAI_API_KEY is set)\n")

    workflow = ResearchWritingWorkflow(max_revision_cycles=2)
    result = workflow.run(topic)

    print_workflow_trace(result.workflow_trace)

    print("\n" + "=" * 60)
    print("FINAL REPORT")
    print("=" * 60)
    print(result.report)
    print()
    print(f"Sources ({len(result.sources)}):")
    for src in result.sources:
        print(f"  • {src}")
    print(f"\nRevisions: {result.revisions}")
    print(f"Total time: {result.elapsed_ms:.0f} ms")


if __name__ == "__main__":
    main()

File: code/go/06-frameworks/langgraph_multi_agent.go
// langgraph_multi_agent.go — LangGraph-equivalent multi-agent state machine in Go.
//
// LangGraph has no official Go port. This file implements the same CONCEPT —
// a directed state machine where specialised agents run as "nodes" and pass
// a shared state object to the next node — using Go-native code.
//
// Workflow (research → fact_check → writer → editor, with revision loop):
//
//	research_node → fact_check_node → writer_node → editor_node
//	                                       ↑_____________|  (if not approved)
//
// The editor decides whether to approve or request another revision.
// If maxRevisions is exceeded, the workflow terminates regardless.
//
// Run:
//
//	go run langgraph_multi_agent.go hybrid_rag_agent.go langgraph_alternative.go \
//	           multi_agent_from_scratch.go framework_comparison.go framework_advisor.go \
//	           framework_extraction.go over_engineering_detector.go langsmith_tracer.go \
//	           autogen_design_team.go crewai_research_crew.go
//
// See: docs/02-the-agent-loop/04-multi-agent-patterns.md
package main

import (
	"context"
	"fmt"
	"strings"
	"time"

	openai "github.com/sashabaranov/go-openai"
)

// ---------------------------------------------------------------------------
// State types
// ---------------------------------------------------------------------------

// LGMAStep records one step in the workflow trace.
type LGMAStep struct {
	Step          string
	Agent         string
	Summary       string
	OutputPreview string
}

// LGMAState is the shared state passed between nodes.
type LGMAState struct {
	Topic          string
	ResearchNotes  string
	VerifiedFacts  string
	DraftReport    string
	EditorFeedback string
	Approved       bool
	RevisionCount  int
	Sources        []string
	WorkflowTrace  []LGMAStep
}

// LGMAResult is returned after the full workflow completes.
type LGMAResult struct {
	Report        string
	Sources       []string
	Revisions     int
	WorkflowTrace []LGMAStep
	ElapsedMs     float64
}

// ---------------------------------------------------------------------------
// Simulated search tools (same data as crewai_research_crew.go but independent)
// ---------------------------------------------------------------------------

var lgmaMockSearchData = map[string]string{
	"AI productivity": "Researchers at Stanford and MIT found AI coding assistants increase output by 20-55% (2023-2024 studies). GitHub Copilot study: 55% faster. NBER: 56% faster for experienced workers.",
	"developer tools": "Stack Overflow 2024: 77% adoption of AI tools. Most used: GitHub Copilot (34%), Cursor (18%), Codeium (12%).",
	"software output": "McKinsey Global Institute (2024): generative AI could automate 30% of software tasks within 3 years.",
	"impact":          "Key risks: over-reliance, security vulnerabilities in AI-generated code, hallucinations in complex logic.",
}

func lgmaSearch(query string) string {
	lower := strings.ToLower(query)
	for key, val := range lgmaMockSearchData {
		if strings.Contains(lower, key) {
			return val
		}
	}
	return fmt.Sprintf("Search: No specific data found for %q. General knowledge used.", query)
}

// ---------------------------------------------------------------------------
// Node functions
// ---------------------------------------------------------------------------

// lgmaResearchNode gathers raw research notes.
func lgmaResearchNode(ctx context.Context, client *openai.Client, state LGMAState) (LGMAState, error) {
	searchResults := lgmaSearch(state.Topic)
	prompt := fmt.Sprintf(
		"You are a research agent. Topic: %q\n\n"+
			"Available data:\n%s\n\n"+
			"Produce comprehensive research notes (200–300 words). Include:\n"+
			"1. Key statistics with sources.\n"+
			"2. Main trends.\n"+
			"3. At least one limitation or counterpoint.\n"+
			"List 3–5 sources at the end.",
		state.Topic, searchResults)

	resp, err := client.CreateChatCompletion(ctx, openai.ChatCompletionRequest{
		Model:     "gpt-4o-mini",
		Messages:  []openai.ChatCompletionMessage{{Role: openai.ChatMessageRoleUser, Content: prompt}},
		MaxTokens: 500,
	})
	if err != nil {
		return state, fmt.Errorf("research node: %w", err)
	}

	notes := resp.Choices[0].Message.Content
	state.ResearchNotes = notes

	// Extract sources mentioned in the notes
	for _, line := range strings.Split(notes, "\n") {
		l := strings.TrimSpace(line)
		if strings.HasPrefix(strings.ToLower(l), "source") || strings.HasPrefix(l, "-") {
			if len(l) > 5 {
				state.Sources = append(state.Sources, l)
			}
		}
	}

	preview := notes
	if len(preview) > 100 {
		preview = preview[:100] + "…"
	}
	state.WorkflowTrace = append(state.WorkflowTrace, LGMAStep{
		Step:          "research",
		Agent:         "ResearchAgent",
		Summary:       fmt.Sprintf("Gathered %d chars of research notes", len(notes)),
		OutputPreview: preview,
	})
	return state, nil
}

// lgmaFactCheckNode verifies claims in the research notes.
func lgmaFactCheckNode(ctx context.Context, client *openai.Client, state LGMAState) (LGMAState, error) {
	prompt := fmt.Sprintf(
		"You are a fact-checking agent. Review these research notes:\n\n%s\n\n"+
			"1. Mark each quantitative claim as VERIFIED, PLAUSIBLE, or UNVERIFIED.\n"+
			"2. Remove or flag any claim you cannot verify.\n"+
			"3. Produce a 'Verified Facts' document with only reliable information.\n"+
			"Be conservative — if in doubt, mark as PLAUSIBLE.",
		state.ResearchNotes)

	resp, err := client.CreateChatCompletion(ctx, openai.ChatCompletionRequest{
		Model:     "gpt-4o-mini",
		Messages:  []openai.ChatCompletionMessage{{Role: openai.ChatMessageRoleUser, Content: prompt}},
		MaxTokens: 500,
	})
	if err != nil {
		return state, fmt.Errorf("fact-check node: %w", err)
	}

	verified := resp.Choices[0].Message.Content
	state.VerifiedFacts = verified

	preview := verified
	if len(preview) > 100 {
		preview = preview[:100] + "…"
	}
	state.WorkflowTrace = append(state.WorkflowTrace, LGMAStep{
		Step:          "fact_check",
		Agent:         "FactCheckAgent",
		Summary:       "Verified claims in research notes",
		OutputPreview: preview,
	})
	return state, nil
}

// lgmaWriterNode drafts the report from verified facts.
func lgmaWriterNode(ctx context.Context, client *openai.Client, state LGMAState) (LGMAState, error) {
	prevFeedback := ""
	if state.EditorFeedback != "" {
		prevFeedback = fmt.Sprintf("\n\nEditor feedback from previous draft:\n%s\n\nPlease address all points above.", state.EditorFeedback)
	}

	prompt := fmt.Sprintf(
		"You are a technical writer. Write a 300–400-word report on: %q\n\n"+
			"Use only these verified facts:\n%s%s\n\n"+
			"Structure: executive summary → key findings → risks → recommendations.\n"+
			"Include inline citations (Author/Org, Year).",
		state.Topic, state.VerifiedFacts, prevFeedback)

	resp, err := client.CreateChatCompletion(ctx, openai.ChatCompletionRequest{
		Model:     "gpt-4o-mini",
		Messages:  []openai.ChatCompletionMessage{{Role: openai.ChatMessageRoleUser, Content: prompt}},
		MaxTokens: 700,
	})
	if err != nil {
		return state, fmt.Errorf("writer node: %w", err)
	}

	draft := resp.Choices[0].Message.Content
	state.DraftReport = draft

	preview := draft
	if len(preview) > 100 {
		preview = preview[:100] + "…"
	}
	label := "Draft"
	if state.RevisionCount > 0 {
		label = fmt.Sprintf("Revision %d", state.RevisionCount)
	}
	state.WorkflowTrace = append(state.WorkflowTrace, LGMAStep{
		Step:          "writer",
		Agent:         "WriterAgent",
		Summary:       fmt.Sprintf("%s written (%d chars)", label, len(draft)),
		OutputPreview: preview,
	})
	return state, nil
}

// lgmaEditorNode reviews the draft and either approves it or requests revisions.
func lgmaEditorNode(ctx context.Context, client *openai.Client, state LGMAState) (LGMAState, error) {
	prompt := fmt.Sprintf(
		"You are a senior editor. Review this draft report:\n\n%s\n\n"+
			"If the report is publication-ready (accurate, well-structured, ≥300 words, has citations), "+
			"respond with exactly: APPROVED\n\n"+
			"Otherwise, provide specific, actionable feedback on what must be improved. "+
			"Be strict — the report must cite sources and have clear structure.",
		state.DraftReport)

	resp, err := client.CreateChatCompletion(ctx, openai.ChatCompletionRequest{
		Model:     "gpt-4o-mini",
		Messages:  []openai.ChatCompletionMessage{{Role: openai.ChatMessageRoleUser, Content: prompt}},
		MaxTokens: 300,
	})
	if err != nil {
		return state, fmt.Errorf("editor node: %w", err)
	}

	feedback := resp.Choices[0].Message.Content
	state.EditorFeedback = feedback
	state.Approved = strings.Contains(strings.ToUpper(feedback), "APPROVED")
	state.RevisionCount++

	summary := "Requested revisions"
	if state.Approved {
		summary = "APPROVED"
	}
	preview := feedback
	if len(preview) > 100 {
		preview = preview[:100] + "…"
	}
	state.WorkflowTrace = append(state.WorkflowTrace, LGMAStep{
		Step:          "editor",
		Agent:         "EditorAgent",
		Summary:       summary,
		OutputPreview: preview,
	})
	return state, nil
}

// ---------------------------------------------------------------------------
// Workflow runner
// ---------------------------------------------------------------------------

// runLGMAWorkflow executes the full research → fact_check → writer ⇄ editor loop.
func runLGMAWorkflow(ctx context.Context, client *openai.Client, topic string, maxRevisions int) (LGMAResult, error) {
	t0 := time.Now()

	state := LGMAState{Topic: topic}

	// Phase 1: Research
	var err error
	state, err = lgmaResearchNode(ctx, client, state)
	if err != nil {
		return LGMAResult{}, err
	}

	// Phase 2: Fact-check
	state, err = lgmaFactCheckNode(ctx, client, state)
	if err != nil {
		return LGMAResult{}, err
	}

	// Phase 3+4: Write ↔ Edit (revision loop)
	for i := 0; i <= maxRevisions; i++ {
		state, err = lgmaWriterNode(ctx, client, state)
		if err != nil {
			return LGMAResult{}, err
		}
		state, err = lgmaEditorNode(ctx, client, state)
		if err != nil {
			return LGMAResult{}, err
		}
		if state.Approved {
			break
		}
	}

	return LGMAResult{
		Report:        state.DraftReport,
		Sources:       state.Sources,
		Revisions:     state.RevisionCount - 1, // first pass is not a revision
		WorkflowTrace: state.WorkflowTrace,
		ElapsedMs:     float64(time.Since(t0).Milliseconds()),
	}, nil
}

// ---------------------------------------------------------------------------
// Reporting
// ---------------------------------------------------------------------------

// lgmaPrintWorkflowTrace prints the workflow steps in a compact table.
func lgmaPrintWorkflowTrace(trace []LGMAStep) {
	fmt.Printf("\n%s\n  WORKFLOW TRACE\n%s\n", strings.Repeat("─", 70), strings.Repeat("─", 70))
	for i, step := range trace {
		fmt.Printf("  [%02d] %-12s  %-18s  %s\n", i+1, step.Step, step.Agent, step.Summary)
		if step.OutputPreview != "" {
			fmt.Printf("       → %s\n", step.OutputPreview)
		}
	}
	fmt.Printf("%s\n", strings.Repeat("─", 70))
}

// ---------------------------------------------------------------------------
// Demo
// ---------------------------------------------------------------------------

// runLangGraphMultiAgentDemo is the demo entry point (no main()).
func runLangGraphMultiAgentDemo(ctx context.Context, client *openai.Client) {
	topics := []string{
		"The impact of AI on software developer productivity",
	}

	for i, topic := range topics {
		fmt.Printf("\n%s\n  LANGGRAPH MULTI-AGENT WORKFLOW — TOPIC %d\n%s\n", strings.Repeat("═", 60), i+1, strings.Repeat("═", 60))
		fmt.Printf("\nTopic: %s\n", topic)
		fmt.Println("Workflow: research → fact_check → writer ⇄ editor")
		fmt.Printf("Max revisions: 2\n\n")

		result, err := runLGMAWorkflow(ctx, client, topic, 2)
		if err != nil {
			fmt.Printf("Error: %v\n", err)
			continue
		}

		lgmaPrintWorkflowTrace(result.WorkflowTrace)

		fmt.Printf("\nRevisions requested: %d\n", result.Revisions)
		fmt.Printf("Total workflow time : %.0fms\n", result.ElapsedMs)
		fmt.Printf("Sources found       : %d\n", len(result.Sources))

		fmt.Printf("\nFinal Report (preview):\n")
		report := result.Report
		if len(report) > 500 {
			report = report[:500] + "\n…(truncated)"
		}
		fmt.Printf("%s\n", report)
	}
}
